File: 09_dijkstras_algorithm/scala/01_dijkstras_algorithm.scala
import scala.collection.mutable

// the graph
val graph = mutable.Map[String, mutable.Map[String, Int]](
  "start" -> mutable.Map("a" -> 6, "b" -> 2),
  "a" -> mutable.Map("fin" -> 1),
  "b" -> mutable.Map("a" -> 3, "fin" -> 5),
  "fin" -> mutable.Map[String, Int]()
)

// the costs table
val infinity = Double.PositiveInfinity
val costs = mutable.Map[String, Double](
  "a" -> 6,
  "b" -> 2,
  "fin" -> infinity
)

// the parents table
val parents = mutable.Map[String, String](
  "a" -> "start",
  "b" -> "start",
  "fin" -> null
)

val processed = mutable.ListBuffer[String]()

def findLowestCostNode(costs: mutable.Map[String, Double], processed: mutable.ListBuffer[String]): Option[String] = {
  var lowestCost = infinity
  var lowestCostNode: Option[String] = None
  // Go through each node.
  for((node, cost) <- costs) {
    // If it's the lowest cost so far and hasn't been processed yet...
    if(cost < lowestCost && !processed.contains(node)) {
      // ... set it as the new lowest-cost node.
      lowestCost = cost
      lowestCostNode = Some(node)
    }
  }
  lowestCostNode
}

// Find the lowest-cost node that you haven't processed yet.
var node = findLowestCostNode(costs, processed)
// If you've processed all the nodes, this while loop is done.
while(node.isDefined) {
  val cost = costs(node.get)
  // Go through all the neighbors of this node.
  val neighbors = graph(node.get)
  for((n, neighborCost) <- neighbors) {
    val newCost = cost + neighborCost
    // If it's cheaper to get to this neighbor by going through this node...
    if(costs(n) > newCost) {
      // ... update the cost for this node.
      costs(n) = newCost
      // This node becomes the new parent for this neighbor.
      parents(n) = node.get
    }
  }
  // Mark the node as processed.
  processed += node.get
  // Find the next node to process, and loop.
  node = findLowestCostNode(costs, processed)
}

println("Cost from the start to each node:")
println(costs)

File: 09_dijkstras_algorithm/clojure/01_dijkstras_algorithm.clj
(ns dijkstras-algorithm)

;; the graph
(def graph {"start" {"a" 6, "b" 2}
            "a" {"fin" 1}
            "b" {"a" 3, "fin" 5}
            "fin" {}})

;; the costs table
(def infinity Double/POSITIVE_INFINITY)
(def costs (atom {"a" 6, "b" 2, "fin" infinity}))

;; the parents table
(def parents (atom {"a" "start", "b" "start", "fin" nil}))

(def processed (atom []))

(defn find-lowest-cost-node [costs processed]
  (let [unprocessed (remove #(contains? processed %) (keys costs))
        costs-seq (map (fn [node] [node (get costs node)]) unprocessed)
        sorted-costs (sort-by second costs-seq)]
    (when (seq sorted-costs)
      (first (first sorted-costs)))))

;; Find the lowest-cost node that you haven't processed yet.
(loop [node (find-lowest-cost-node @costs @processed)]
  ;; If you've processed all the nodes, this loop is done.
  (when node
    (let [cost (get @costs node)
          ;; Go through all the neighbors of this node.
          neighbors (get graph node {})]
      (doseq [[n neighbor-cost] neighbors]
        (let [new-cost (+ cost neighbor-cost)]
          ;; If it's cheaper to get to this neighbor by going through this node...
          (when (> (get @costs n infinity) new-cost)
            ;; ... update the cost for this node.
            (swap! costs assoc n new-cost)
            ;; This node becomes the new parent for this neighbor.
            (swap! parents assoc n node))))
      ;; Mark the node as processed.
      (swap! processed conj node)
      ;; Find the next node to process, and loop.
      (recur (find-lowest-cost-node @costs @processed)))))

(println "Cost from the start to each node:")
(println @costs)

